src/Data/Variant.purs

```diff
@@ -73,6 +79,70 @@ on p f g r =
   coerceR ∷ Variant r2 → Variant r1
   coerceR = unsafeCoerce
 
+-- | Type class that matches a row for a `record` that will eliminate a row for
+-- | a `variant`, producing a `result` of the specified type. Just a wrapper for
+-- | `VariantMatch` to convert `RowToList` and vice versa.
+class VariantRecordElim
+    (variant ∷ # Type)
+    (record ∷ # Type)
+    result
+  | variant result → record
+  , record → variant result
+instance variantRecordElim
+  ∷ ( RowToList variant vlist
+    , RowToList record rlist
+    , VariantMatch vlist rlist result
+    , ListToRow vlist variant
+    , ListToRow rlist record )
+  ⇒ VariantRecordElim variant record result
+
+-- | Checks that a `RowList` matches the argument to be given to the function
+-- | in the other `RowList` with the same label, such that it will produce the
+-- | result type.
+class VariantMatch
+    (vlist ∷ R.RowList)
+    (rlist ∷ R.RowList)
+    result
+  | vlist result → rlist
+  , rlist → vlist result
+instance variantMatchNil
+  ∷ VariantMatch R.Nil R.Nil r
+instance variantMatchCons
+  ∷ VariantMatch v r res
+  ⇒ VariantMatch (R.Cons sym a v) (R.Cons sym (a → res) r) res
+
+-- | Eliminate a `variant` with a `record` containing methods to handle each
+-- | case and produce a unified `result`.
+-- |
+-- | This means that if `variant` contains a row of type `a`, a row with the
+-- | same label must have type `a -> b` in `record`, where `b` is the same
+-- | `result` type for every row of `record`.
+-- |
+-- | Methods in `record` cannot have any constraints, due to limitations in
+-- | PureScript's type system. This means that class methods must be given a
+-- | concrete type, as in `{ foo: show :: Int -> String }`. Other functions
+-- | using `forall` will still work if they do not contain constraints, but note
+-- | that even `id` must be given a type annotation to be used in this way,
+-- | since it normally depends on `Category`: use `id :: forall a. a -> a` not
+-- | `id :: forall a c. Category c => c a a`.
+elim
```

Comment:
I'm crossing wires a bit :) The signature for `foldList` I gave is the fold for lists, whether or not it's represented with `Variant` or data constructors or Church-encoding. It's not possible to define a definition of `List` with `Variant` without a newtype or `VariantF` + `Mu`. If I were to use `VariantF`, it might be:

```purescript
type ListF a f = VariantF (nil :: FProxy (Const Unit), cons :: FProxy (Tuple a f)))
type List a = Mu (ListF a)
```

In which case, the proposed record eliminator would be the actual fold when run with `cata` (a generalized folding operation).